File: src/main/java/nox/compilation/BuildPropertiesCreationAction.java
/*
 * Copyright (c) Oleg Sklyar 2017. License: MIT
 */
package nox.compilation;

import com.google.common.collect.Lists;
import org.apache.commons.lang3.StringUtils;
import org.gradle.api.GradleException;
import org.gradle.api.Project;
import org.gradle.api.file.SourceDirectorySet;
import org.gradle.api.internal.plugins.ExtensionContainerInternal;
import org.gradle.api.internal.project.ProjectInternal;
import org.gradle.api.internal.tasks.TaskContainerInternal;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSetContainer;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;


public class BuildPropertiesCreationAction {

	private static final String BUILD_PROPERTIES = "build.properties";

	private final File projectDir;

	private final JavaPluginConvention javaConv;

	private final OSGiExt.BuildProperties buildProps;

	public BuildPropertiesCreationAction(Project target) {
		ProjectInternal project = (ProjectInternal) target;

		TaskContainerInternal tasks = project.getTasks();
		ExtensionContainerInternal ext = project.getExtensions();

		projectDir = project.getProjectDir();
		javaConv = project.getConvention().getPlugin(JavaPluginConvention.class);
		buildProps = ext.findByType(OSGiExt.class).buildProperties;

		tasks.getByName("jar").doFirst(task -> create());
		tasks.getByName("clean").doFirst(task -> clean());
	}

	private void create() {
		List<String> lines = Lists.newArrayList();
		List<String> javaSources = getSources(buildProps.sources, ss -> ss.getByName("main").getJava());
		javaSources.addAll(getSources(Lists.newArrayList(), ss -> ss.getByName("main").getResources()));
		if (!javaSources.isEmpty()) {
			lines.add("source.. = " + StringUtils.join(javaSources, ","));
		}
		List<String> bins = buildProps.binincludes.stream()
			.filter(path -> "META-INF/".equals(path) || new File(projectDir, path).exists())
			.collect(Collectors.toList());
		if (!bins.isEmpty()) {
			lines.add("bin.includes = " + StringUtils.join(bins, ","));
		}
		if (!javaSources.isEmpty()) {
			lines.add("output.. = " + (buildProps.outputs.isEmpty() ? "bin/" : StringUtils.join(buildProps.outputs, ",")));
		}
		for (Map.Entry<String, String> entry: buildProps.instructions.entrySet()) {
			lines.add(String.format("%s=%s", entry.getKey(), entry.getValue()));
		}

		try {
			Files.write(Paths.get(projectDir.getAbsolutePath(), BUILD_PROPERTIES), lines, Charset.forName("UTF-8"));
		} catch (IOException ex) {
			throw new GradleException("Failed to create build.properties for an OSGi bundle", ex);
		}
	}

	private List<String> getSources(List<String> preconfigured,  Function<SourceSetContainer, SourceDirectorySet> sourceExtractor) {
		if (!preconfigured.isEmpty()) {
			return preconfigured;
		}
		Set<File> sourceEntries = sourceExtractor.apply(javaConv.getSourceSets()).getSrcDirs();
		return sourceEntries.stream()
			.filter(se -> se.exists())
			.map(se -> {
				String element = se.getAbsolutePath().replace(projectDir.getAbsolutePath() + "/", "");
				return element + (!element.endsWith("/") ? "/" : "");
			})
			.collect(Collectors.toList());
	}

	private void clean() {
		new File(projectDir, BUILD_PROPERTIES).delete();
	}
}

File: src/main/kotlin/nox/compilation/BuildPropertiesCreationAction.kt
/*
 * Copyright (c) Oleg Sklyar 2017. License: MIT
 */
package nox.compilation

import org.gradle.api.GradleException
import org.gradle.api.Project
import org.gradle.api.file.SourceDirectorySet
import org.gradle.api.internal.project.ProjectInternal
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.api.tasks.SourceSetContainer
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Paths


class BuildPropertiesCreationAction(target: Project) {

	private val projectDir: File

	private val javaConv: JavaPluginConvention

	private val buildProps: OSGiExt.BuildProperties

	init {
		val project = target as ProjectInternal

		val tasks = project.tasks
		val ext = project.extensions

		projectDir = project.projectDir
		javaConv = project.convention.getPlugin(JavaPluginConvention::class.java)
		buildProps = ext.findByType(OSGiExt::class.java).buildProperties

		tasks.getByName("jar").doFirst { _ -> create() }
		tasks.getByName("clean").doFirst { _ -> clean() }
	}

	internal interface SourceExtractor {
		fun extract(sourceSets: SourceSetContainer): SourceDirectorySet
	}

	private val javaExtractor = object : SourceExtractor {
		override fun extract(sourceSets: SourceSetContainer): SourceDirectorySet {
			return sourceSets.getByName("main").java
		}
	}

	private val resourceExtractor = object : SourceExtractor {
		override fun extract(sourceSets: SourceSetContainer): SourceDirectorySet {
			return sourceSets.getByName("main").resources
		}
	}
	private fun create() {
		val lines = mutableListOf<String>()
		var javaSources = getSources(javaExtractor)
		javaSources.addAll(buildProps.sources)
		javaSources = javaSources.filter { File(projectDir, it).exists() }.toMutableList()
		if (javaSources.isNotEmpty()) {
			lines.add("source.. = " + javaSources.joinToString(","))
			lines.add("output.. = " + if (buildProps.outputs.isEmpty()) "bin/" else buildProps.outputs.joinToString(","))
		}
		var bins = getSources(resourceExtractor)
		bins.addAll(buildProps.binincludes)
		bins = bins.filter { it == "META-INF/" || it == "." || File(projectDir, it).exists() }.toMutableList()
		if (bins.isNotEmpty()) {
			lines.add("bin.includes = " + bins.joinToString(","))
		}
		for ((key, value) in buildProps.instructions) {
			lines.add("$key=$value")
		}

		try {
			Files.write(Paths.get(projectDir.absolutePath, BUILD_PROPERTIES), lines, Charset.forName("UTF-8"))
		} catch (ex: IOException) {
			throw GradleException("Failed to create build.properties for an OSGi bundle", ex)
		}

	}

	private fun getSources(sourceExtractor: SourceExtractor): MutableList<String> {
		val sourceEntries = sourceExtractor.extract(javaConv.sourceSets).srcDirs
		return sourceEntries
			.filter { it.exists() }
			.map {
				var element = it.absolutePath.replace(projectDir.absolutePath + "/", "").replace(projectDir.absolutePath + "\\", "")
				element += if (!element.endsWith("/")) "/" else ""
				element.replace("\\", "/")
			}.toMutableList()
	}

	private fun clean() {
		File(projectDir, BUILD_PROPERTIES).delete()
	}

	companion object {

		private val BUILD_PROPERTIES = "build.properties"
	}
}
